Zend/zend_operators.c

```diff
@@ -1840,106 +1840,138 @@ ZEND_API zend_result ZEND_FASTCALL shift_right_function(zval *result, zval *op1,
-	 			if (Z_TYPE_P(op2) == IS_STRING) break;
-	 		}
-			ZEND_TRY_BINARY_OP2_OBJECT_OPERATION(ZEND_CONCAT);
-			ZVAL_STR(&op2_copy, zval_get_string_func(op2));
-			if (UNEXPECTED(EG(exception))) {
-				zval_ptr_dtor_str(&op1_copy);
-				zval_ptr_dtor_str(&op2_copy);
-				if (orig_op1 != result) {
-					ZVAL_UNDEF(result);
+		if (!op2_string) {
+			if (EXPECTED(Z_TYPE_P(op2) == IS_STRING)) {
+				op2_string = Z_STR_P(op2);
+			} else {
+				if (Z_ISREF_P(op2)) {
+					op2 = Z_REFVAL_P(op2);
+					if (Z_TYPE_P(op2) == IS_STRING) {
+						op2_string = Z_STR_P(op2);
+						break;
+					}
 				}
-				return FAILURE;
+				/* hold an additional reference because a userland function could free this */
+				if (!op1_is_non_string_or_copy) {
+					op1_string = zend_string_copy(op1_string);
+					op1_is_non_string_or_copy = true;
+				}
+				ZEND_TRY_BINARY_OP2_OBJECT_OPERATION(ZEND_CONCAT);
+				op2_string = zval_get_string_func(op2);
+				if (UNEXPECTED(EG(exception))) {
+					zend_string_release(op1_string);
+					zend_string_release(op2_string);
+					if (orig_op1 != result) {
+						ZVAL_UNDEF(result);
+					}
+					return FAILURE;
+				}
+				op2_is_non_string_or_copy = true;
 			}
-			op2 = &op2_copy;
 		}
 	} while (0);
 
-	if (UNEXPECTED(Z_STRLEN_P(op1) == 0)) {
-		if (EXPECTED(result != op2)) {
+	if (UNEXPECTED(ZSTR_LEN(op1_string) == 0)) {
+		if (EXPECTED(op2_is_non_string_or_copy || result != op2)) {
 			if (result == orig_op1) {
 				i_zval_ptr_dtor(result);
 			}
-			ZVAL_COPY(result, op2);
+			if (op2_is_non_string_or_copy) {
+				ZVAL_STR(result, op2_string);
+				op2_is_non_string_or_copy = false;
+			} else {
+				ZVAL_STR_COPY(result, op2_string);
+			}
 		}
-	} else if (UNEXPECTED(Z_STRLEN_P(op2) == 0)) {
-		if (EXPECTED(result != op1)) {
+	} else if (UNEXPECTED(ZSTR_LEN(op2_string) == 0)) {
+		if (EXPECTED(op1_is_non_string_or_copy || result != op1)) {
 			if (result == orig_op1) {
 				i_zval_ptr_dtor(result);
 			}
-			ZVAL_COPY(result, op1);
+			if (op1_is_non_string_or_copy) {
+				ZVAL_STR(result, op1_string);
+				op1_is_non_string_or_copy = false;
+			} else {
+				ZVAL_STR_COPY(result, op1_string);
+			}
 		}
 	} else {
-		size_t op1_len = Z_STRLEN_P(op1);
-		size_t op2_len = Z_STRLEN_P(op2);
+		size_t op1_len = ZSTR_LEN(op1_string);
+		size_t op2_len = ZSTR_LEN(op2_string);
 		size_t result_len = op1_len + op2_len;
 		zend_string *result_str;
 
 		if (UNEXPECTED(op1_len > ZSTR_MAX_LEN - op2_len)) {
+			if (op1_is_non_string_or_copy) zend_string_release(op1_string);
+			if (op2_is_non_string_or_copy) zend_string_release(op2_string);
 			zend_throw_error(NULL, "String size overflow");
-			zval_ptr_dtor_str(&op1_copy);
-			zval_ptr_dtor_str(&op2_copy);
 			if (orig_op1 != result) {
 				ZVAL_UNDEF(result);
 			}
 			return FAILURE;
 		}
 
-		if (result == op1 && Z_REFCOUNTED_P(result)) {
+		if (result == op1) {
+			if (op1_is_non_string_or_copy) {
+				i_zval_ptr_dtor(result);
+				op1_is_non_string_or_copy = false;
```

Comment:
Given how these values are used, maybe a better name would be `free_op1`.

Reply:
That's indeed better. Actually I'd prefer `free_op1_string` to make it clear it's about the string, not about the op.